### PR TITLE
Assignee of issue/PR receive notifications for new comments

### LINKED ISSUE #391: Assignee of issue/PR should receive notifications for new comments
Currently only the creator of a PR/issue receives a notification when a new comment is made on the issue/PR. Assignees should also receive these notifications, but they should not receive two notifications if they are also the creator of the issue/PR.

### PR DESCRIPTION
**Pull request by sanjaydemansol**
#### Testing notes
- Add user/users as assignee to an issue
- Comment on the issue
- Assignee's receive direct message from Bot on Mattermost

#### Summary
Fix #391 

#### Ticket Link
#391

### CODE REVIEW

File: server/plugin/webhook.go [hunk 1]
```diff
@@ -893,6 +894,68 @@ func (p *Plugin) handleCommentAuthorNotification(event *github.IssueCommentEvent
...
+		if authorUserID == "" {
+			return
+		}
+
+		assigneeID := p.getGitHubToUserIDMapping(assignee.GetLogin())
+		if assigneeID == "" {
+			return
+		}
+
+		if event.GetRepo().GetPrivate() && !p.permissionToRepo(authorUserID, event.GetRepo().GetFullName()) {
```
Comment (hanzei) [inline server/plugin/webhook.go:924]: Why does this check if the author has access, given that the event is send to assignee?

File: server/plugin/webhook.go [hunk 2]
```diff
@@ -893,6 +894,68 @@ func (p *Plugin) handleCommentAuthorNotification(event *github.IssueCommentEvent
...
+
+		splitURL := strings.Split(event.GetIssue().GetHTMLURL(), "/")
+		if len(splitURL) < 2 {
+			return
+		}
+
+		var templateName string
+		switch splitURL[len(splitURL)-2] {
+		case "pull":
+			templateName = "commentAuthorPullRequestNotification"
```
Comment (hanzei) [inline server/plugin/webhook.go:936]: Should there be different message? It's not really "your pull request" (https://github.com/mattermost/mattermost-plugin-github/blob/41f3bf52bd4b98e416aafccf96878aeec773da77/server/plugin/template.go#L279)

**Review by hanzei — CHANGES_REQUESTED** [review]
Nice work. A few question below

**Comment by dipak-demansol** [comment]
@sanjaydemansol there is a issue, pls check below steps to regenerate it.

1) i created repo, in that 2nd user created issue.
2) 2nd user assign that issue to him self and to the 3rd user.
3) now repo owner as i comment on that issue then 2nd & 3rd user is not getting the notification but if i remove any 1 assignee and add comment again then single user is getting the notification.

File: server/plugin/webhook.go (hunk 1 above)
Comment (mohit-deshwal) [reply server/plugin/webhook.go:924]: Okay, will Remove it.

Comment (sanjaydemansol) [reply server/plugin/webhook.go:924]: Okay, will remove it.

File: server/plugin/webhook.go (hunk 2 above)
Comment (sanjaydemansol) [reply server/plugin/webhook.go:936]: yeah, used template of author. Any recommendations for template?

Comment (hanzei) [reply server/plugin/webhook.go:936]: No specific ones. I'm happy to review your proposal.

**Comment by dipak-demansol** [comment]
@sanjaydemansol  there are more test cases where i'm facing the issue, i'm attaching the screenshot of the test case.
below colour indicates 
red = user1(I)
magenta = 2nd user
blue = 3rd user
bold black color = Fail(issue)
Screenshot from 2021-10-04 15-23-26

File: server/plugin/webhook.go [hunk 3]
```diff
@@ -893,6 +894,68 @@ func (p *Plugin) handleCommentAuthorNotification(event *github.IssueCommentEvent
...
+			return
+		}
+
+		authorUserID := p.getGitHubToUserIDMapping(author)
+		if authorUserID == "" {
+			return
+		}
+
+		assigneeID := p.getGitHubToUserIDMapping(assignee.GetLogin())
+		if assigneeID == "" {
```
Comment (maisnamrajusingh) [inline server/plugin/webhook.go:920]: @hanzei  is this an acceptable way of testing for empty strings or do you have any suggestions ?

**Review by maisnamrajusingh — APPROVED** [review]
Apart from the question above I think this is ok by me.

File: server/plugin/webhook.go (hunk 3 above)
Comment (hanzei) [reply server/plugin/webhook.go:920]: Testing for `""` is the idiomatic way to test that in go.

**Comment by sanjaydemansol** [comment]
> @sanjaydemansol there are more test cases where i'm facing the issue, i'm attaching the screenshot of the test case. below colour indicates red = user1(I) magenta = 2nd user blue = 3rd user bold black color = Fail(issue) Screenshot from 2021-10-04 15-23-26

@dipak-demansol Thanks for your efforts for such a detailed report, turns out I used `return` statement instead of `continue` which lead to these issues. Fixed.

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking on https://github.com/mattermost/mattermost-plugin-github/pull/485#discussion_r736129225

File: server/plugin/webhook.go [hunk 4]
```diff
@@ -893,6 +894,58 @@ func (p *Plugin) handleCommentAuthorNotification(event *github.IssueCommentEvent
...
+			continue
+		}
+
+		if author == assignee.GetLogin() {
+			continue
+		}
+		if event.Sender.GetLogin() == assignee.GetLogin() {
+			continue
+		}
+
```
[the remark this answers is not in the document]
Comment (sanjaydemansol) [reply server/plugin/webhook.go:929]: added check for the suggested case. Thanks.

**Review by hanzei — APPROVED** [review]

**Pull request merged**